storage/table_entities.go

```diff
@@ -0,0 +1,379 @@
+
+	headers["Content-Length"] = fmt.Sprintf("%d", buf.Len())
+
+	var err error
+	var resp *odataResponse
+
+	if updateInsteadOfMerge {
+		resp, err = c.client.execTable("PUT", uri, headers, buf)
+	} else {
+		resp, err = c.client.execTable("MERGE", uri, headers, buf)
+	}
+
+	if err != nil {
+		return err
+	}
+	defer resp.body.Close()
+
+	if err := checkRespCode(resp.statusCode, []int{http.StatusCreated}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// DeleteEntityWithoutCheck deletes the entity matching by
+// PartitionKey and RowKey. There is no check on IfMatch
+// parameter so the entity is always deleted.
+// The function fails if there is no entity
+// with the same PartitionKey and RowKey in the table.
+func (c *TableServiceClient) DeleteEntityWithoutCheck(table AzureTable, entity TableEntity) error {
+	return c.DeleteEntity(table, entity, "*")
+}
+
+// DeleteEntity deletes the entity matching by
+// PartitionKey, RowKey and ifMatch field.
+// The function fails if there is no entity
+// with the same PartitionKey and RowKey in the table or
+// the ifMatch is different.
+func (c *TableServiceClient) DeleteEntity(table AzureTable, entity TableEntity, ifMatch string) error {
+	uri := c.client.getEndpoint(tableServiceName, pathForTable(table), url.Values{})
+	uri += fmt.Sprintf("(PartitionKey='%s',RowKey='%s')", url.QueryEscape(entity.PartitionKey()), url.QueryEscape(entity.RowKey()))
+
+	headers := c.getStandardHeaders()
+
+	headers["Content-Length"] = "0"
+	headers["If-Match"] = ifMatch
+
+	resp, err := c.client.execTable("DELETE", uri, headers, nil)
+
+	if err != nil {
+		return err
+	}
+	defer resp.body.Close()
+
+	if err := checkRespCode(resp.statusCode, []int{http.StatusNoContent}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// InsertOrReplaceEntity inserts an entity in the specified table
+// or replaced the existing one.
+func (c *TableServiceClient) InsertOrReplaceEntity(table AzureTable, entity TableEntity) error {
+	return c.insertOrReplaceOrMergeEntity(table, entity, false)
+}
+
+// InsertOrMergeEntity inserts an entity in the specified table
+// or merges the existing one.
+func (c *TableServiceClient) InsertOrMergeEntity(table AzureTable, entity TableEntity) error {
+	return c.insertOrReplaceOrMergeEntity(table, entity, true)
+}
+
+func (c *TableServiceClient) insertOrReplaceOrMergeEntity(table AzureTable, entity TableEntity, mergeInsteadOfReplace bool) error {
+	uri := c.client.getEndpoint(tableServiceName, pathForTable(table), url.Values{})
+	uri += fmt.Sprintf("(PartitionKey='%s',RowKey='%s')", url.QueryEscape(entity.PartitionKey()), url.QueryEscape(entity.RowKey()))
+
+	headers := c.getStandardHeaders()
+
+	buf := new(bytes.Buffer)
+
+	if err := injectPartitionAndRowKeys(entity, buf); err != nil {
+		return err
+	}
+
+	//	log.Printf("request.body == %s", string(buf.Bytes()))
+
+	headers["Content-Length"] = fmt.Sprintf("%d", buf.Len())
+
+	var err error
+	var resp *odataResponse
+
+	if mergeInsteadOfReplace {
```

Comment:
Yes your logic is sound. I've change the code to follow your advice:

```go
func (c *TableServiceClient) execTable(table AzureTable, entity TableEntity, specifyKeysInUrl bool, method string) (int, error) {
	uri := c.client.getEndpoint(tableServiceName, pathForTable(table), url.Values{})
	if specifyKeysInUrl {
		uri += fmt.Sprintf("(PartitionKey='%s',RowKey='%s')", url.QueryEscape(entity.PartitionKey()), url.QueryEscape(entity.RowKey()))
	}

	headers := c.getStandardHeaders()

	var buf bytes.Buffer

	if err := injectPartitionAndRowKeys(entity, &buf); err != nil {
		return 0, err
	}

	headers["Content-Length"] = fmt.Sprintf("%d", buf.Len())

	var err error
	var resp *odataResponse

	resp, err = c.client.execTable(method, uri, headers, &buf)

	if err != nil {
		return 0, err
	}

	defer resp.body.Close()

	return resp.statusCode, nil
}
```

The insert/whaever methods became:

```go
func (c *TableServiceClient) InsertEntity(table AzureTable, entity TableEntity) error {
	if sc, err := c.execTable(table, entity, false, "POST"); err != nil {
		return checkRespCode(sc, []int{http.StatusCreated})
	} else {
		return err
	}
}

func (c *TableServiceClient) UpdateEntity(table AzureTable, entity TableEntity) error {
	if sc, err := c.execTable(table, entity, true, "PUT"); err != nil {
		return checkRespCode(sc, []int{http.StatusNoContent})
	} else {
		return err
	}
}
```

And so on :)